### PR TITLE
Bring SpanExtensions/ReadOnlySpanExtensions up to CoreClr parity.

### PR DESCRIPTION
**Pull request by AtsushiKan**
This is a necessary step to removing Span from corefxlab.
corefxlab cannot implement these from outside the surface area.

### CODE REVIEW

**Comment by AtsushiKan** [comment]
cc @davidfowl 
cc @KrzysztofCwalina

**Comment by AtsushiKan** [comment]
cc @davidfowl
cc @KrzysztofCwalina

Ping... the sooner this gets merged, the sooner Span is out of corefxlab...

File: src/System.Memory/src/System/ReadOnlySpanExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    /// </summary>
+    public static class ReadOnlySpanExtensions
+    {
+        /// <summary>
+        /// Creates a new readonly span over the portion of the target string.
+        /// </summary>
+        /// <param name="text">The target string.</param>
+        /// <exception cref="System.ArgumentNullException">Thrown when <paramref name="text"/> is null.</exception>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static ReadOnlySpan<char> Slice(this string text)
```
Comment (KrzysztofCwalina) [inline src/System.Memory/src/System/ReadOnlySpanExtensions.cs]: Why does this method need to be in corefx? Could we still keep it as an extension in corfxlab?

File: src/System.Memory/src/System/ReadOnlySpanExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+            unsafe
+            {
+                byte* byteOffset = ((byte*)StringAdjustment) + (uint)(start * sizeof(char));
+                return new ReadOnlySpan<char>(Unsafe.As<Pinnable<char>>(text), (IntPtr)byteOffset, length);
+            }
+        }
+
+        private static readonly IntPtr StringAdjustment = MeasureStringAdjustment();
+
+        private static IntPtr MeasureStringAdjustment()
```
Comment (AtsushiKan) [inline src/System.Memory/src/System/ReadOnlySpanExtensions.cs]: I thought that's what I was doing...

**Comment by KrzysztofCwalina** [comment]
Thanks! Just a couple of small comments. Otherwise looks great!

File: src/System.Memory/src/System/SpanExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        /// </summary>
+        /// <param name="source">The source slice, of type <typeparamref name="T"/>.</param>
+        /// <exception cref="System.ArgumentException">
+        /// Thrown when <typeparamref name="T"/> contains pointers.
+        /// </exception>
+        /// <exception cref="System.OverflowException">
+        /// Thrown if the Length property of the new Span would exceed Int32.MaxValue.
+        /// </exception>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static Span<byte> AsBytes<T>(this Span<T> source)
```
Comment (AtsushiKan) [inline src/System.Memory/src/System/SpanExtensions.cs]: I'm following CoreClr's lead here.

File: src/System.Memory/src/System/ReadOnlySpanExtensions.cs (hunk 1 above)
Comment (AtsushiKan) [inline src/System.Memory/src/System/ReadOnlySpanExtensions.cs]: This isn't implementable using the public surface area (especially w/ corefxlab still on C#6).

DangerousCreate() takes a "ref T" - corefxlab wanted it it to take an IntPtr.

Comment (KrzysztofCwalina) [inline src/System.Memory/src/System/ReadOnlySpanExtensions.cs]: ah, makes sense.

File: src/System.Memory/src/System/ReadOnlySpanExtensions.cs (hunk 2 above)
Comment (KrzysztofCwalina) [inline src/System.Memory/src/System/ReadOnlySpanExtensions.cs]: Ah, sorry. I did not notice the this is called once. Why don't you simply do this work in the static ctor?

Comment (KrzysztofCwalina) [inline src/System.Memory/src/System/ReadOnlySpanExtensions.cs]: I know. I was suggesting that the following makes it clearer (and more reliable) that the initialization static won't be executed twice:
```c#
static IntPtf s_stringAdjustement;
spatic SpanExtensions(){
   s_stringAdjustement = ...
}
```

**Pull request merged**